subprojects/core/src/main/java/org/gradle/initialization/BuildOperatingFiringTaskExecutionPreparer.java

```diff
@@ -67,17 +76,52 @@ public void run(BuildOperationContext buildOperationContext) {
                     return toTaskPaths(taskGraph.getFilteredTasks());
                 }
 
-                private List<String> toTaskPaths(Set<Task> tasks) {
-                    return ImmutableSortedSet.copyOf(Collections2.transform(tasks, new Function<Task, String>() {
-                        @Override
-                        public String apply(Task task) {
-                            return task.getPath();
+                @Override
+                public List<PlannedTask> getTaskPlan() {
+                    List<Node> scheduledWork = taskGraph.getScheduledWork();
+                    Set<DefaultPlannedTask> taskPlan = toPlannedTasks(scheduledWork);
+                    return new ArrayList(taskPlan);
+                }
+
+                private Set<DefaultPlannedTask> toPlannedTasks(List<Node> scheduledWork) {
+                    Set<DefaultPlannedTask> plannedTasks = new LinkedHashSet<>();
+                    for (Node node : scheduledWork) {
+                        if (node instanceof TaskNode) {
+                            TaskNode taskNode = (TaskNode) node;
+                            plannedTasks.add(toPlannedTask(taskNode));
                         }
-                    })).asList();
+                    }
+
+                    return plannedTasks;
+                }
+
+                private DefaultPlannedTask toPlannedTask(TaskNode taskNode) {
+                    TaskIdentity<?> taskIdentity = taskNode.getTask().getTaskIdentity();
+                    return new DefaultPlannedTask(taskIdentity,
+                        transformToIdentities(taskNode.getDependencySuccessors(), node -> node.getDependencySuccessors()),
+                        transformToIdentities(taskNode.getMustSuccessors(), node -> Collections.emptySet()),
+                        transformToIdentities(taskNode.getShouldSuccessors(), node -> Collections.emptySet()),
+                        transformToIdentities(taskNode.getFinalizers(), node -> Collections.emptySet()));
+                }
+
+                private List<TaskIdentity> transformToIdentities(Set<Node> nodes, Function<Node, Set<Node>> nestedNodesResolver) {
+                    return toOnlyTaskNodes(nodes, nestedNodesResolver).stream().map(node -> ((TaskNode) node).getTask().getTaskIdentity())
+                        .collect(Collectors.toList());
+                }
+
+                private List<String> toTaskPaths(Set<Task> tasks) {
+                    return ImmutableSortedSet.copyOf(Collections2.transform(tasks, (Function<Task, String>) task -> task.getPath())).asList();
                 }
             });
         }
 
+        private List<Node> toOnlyTaskNodes(Set<Node> nodes, Function<Node, Set<Node>> nestedNodesResolver) {
+            return nodes.stream()
+                .flatMap(node -> node instanceof TaskNode ? Stream.of(node) : nestedNodesResolver.apply(node).stream())
```

Comment:
This only works for one level of dependencies. If a node depends on a node which depends on a node which finally depends on a task node, then the dependency would not be captured here.

This can for example happen for a chain of artifact transforms, where the transformed artifact is produced by a task or one of the parameters of the transforms is produced by a task.

Reply:
fixed + added test coverage for that scenario